### ISSUE TITLE
Redesign the Interpreter

### ISSUE THREAD

**Issue by Skrinikov** [insufficiently-caffeinated/caffeine#202]
The current implementation of the interpreter is limiting our ability to insert sub operations into the symbolic execution process. We need something which would allow us to inject extra logic into the executor so that we can optimize the program/guide the program using some fuzzer. 
The idea is to have have an executor class which drives the overall program as opposed to just running the symbolic executor as we get the LLVM IR.

Executor Requirements:
- Be able to do some work before we start to do the symbolic execution process.
- be able to do some work after a context exits
   - e.g. after some branch is fully traversed, before getting the next context.
- After the symbolic engine ran.

**Comment by Skrinikov** [comment]
#196 Kinda related

**Comment by Skrinikov** [comment]
[This might help you decide how to design](https://www.usenix.org/system/files/conference/atc13/atc13-bugrara.pdf )

**Comment by Phantomical** [comment]
Closing as I think this is done. If we need to make more changes that can be done under a new issue.
